src/snmpd/patch-5.7.3+dfsg/0009-fix-compilation-in-buster-env.patch

```diff
@@ -0,0 +1,21 @@
+diff --git a/apps/snmptrapd_sql.c b/apps/snmptrapd_sql.c
+index 671e542..68c4b8c 100644
+--- a/apps/snmptrapd_sql.c
++++ b/apps/snmptrapd_sql.c
+@@ -438,13 +438,13 @@ netsnmp_mysql_init(void)
+     }
+ 
+ #ifdef HAVE_BROKEN_LIBMYSQLCLIENT
+-    my_init();
++    //my_init();
```

Comment:
>//my_init(); [](start = 5, length = 12)

can you fix the code, instead of comment out?

Reply:
Done

Reply:
Could you help us understand the details such as:
1. what is the compilation error message?
2. what is the value of HAVE_BROKEN_LIBMYSQLCLIENT?

Also applied to load_defaults? We would like to learn from your investigation.

---
In reply to: [419745538](https://github.com/Azure/sonic-buildimage/pull/4529#discussion_r419745538) [](ancestors = 419745538)

Reply:
```
libtool: link: gcc -DNETSNMP_ENABLE_IPV6 -fno-strict-aliasing -g -O2 -fdebug-prefix-map=/sonic/src/snmpd/net-snmp-5.7.3+dfsg=. -specs=/usr/share/dpkg/no-pie-compile.specs -fstack-protector-strong -Wformat -Werror=format-security -DNETSNMP_USE_INLINE -Ulinux -Dlinux=linux -D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fwrapv -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/lib/x86_64-linux-gnu/perl/5.28/CORE -o .libs/snmptrapd .libs/snmptrapd.o -specs=/usr/share/dpkg/no-pie-link.specs -Wl,-z -Wl,relro -Wl,-z -Wl,now -Wl,-E  ./.libs/libnetsnmptrapd.so ../agent/.libs/libnetsnmpmibs.so ../agent/.libs/libnetsnmpagent.so ../snmplib/.libs/libnetsnmp.so -L/usr/lib/x86_64-linux-gnu/ -lmariadb -lnl-3 -lm -lwrap
/usr/bin/ld: ./.libs/libnetsnmptrapd.so: undefined reference to `load_defaults'
/usr/bin/ld: ./.libs/libnetsnmptrapd.so: undefined reference to `my_init'
collect2: error: ld returned 1 exit status
make[5]: *** [Makefile:297: snmptrapd] Error 1
make[5]: Leaving directory '/sonic/src/snmpd/net-snmp-5.7.3+dfsg/apps'
make[4]: *** [Makefile:657: subdirs] Error 1
make[4]: Leaving directory '/sonic/src/snmpd/net-snmp-5.7.3+dfsg'
dh_auto_build: make -j1 returned exit code 2
make[3]: *** [debian/rules:65: override_dh_auto_build] Error 2
make[3]: Leaving directory '/sonic/src/snmpd/net-snmp-5.7.3+dfsg'
make[2]: *** [debian/rules:41: binary] Error 2
make[2]: Leaving directory '/sonic/src/snmpd/net-snmp-5.7.3+dfsg'
make[1]: *** [Makefile:19: /sonic/target/debs/buster/libsnmp-base_5.7.3+dfsg-1.5_all.deb] Error 2
```

The definitions for "load_defaults"  and "my_init" dont exist.

Reply:
Thanks for the details! I compared it with a stretch build log [https://sonic-jenkins.westus2.cloudapp.azure.com/job/mellanox/job/buildimage-mlnx-all/lastSuccessfulBuild/artifact/target/debs/stretch/libsnmp-base_5.7.3+dfsg-1.5_all.deb.log](https://sonic-jenkins.westus2.cloudapp.azure.com/job/mellanox/job/buildimage-mlnx-all/lastSuccessfulBuild/artifact/target/debs/stretch/libsnmp-base_5.7.3+dfsg-1.5_all.deb.log)
I think it is due to the missing package of libmariadbclient18 in buster. Could you help investigate could we use stretch version? [https://packages.debian.org/stretch/libmariadbclient18](https://packages.debian.org/stretch/libmariadbclient18)

As the macro indicate something is broken in the auto detected LIBMYSQLCLIENT. So the better way is to find a good one.

---
In reply to: [420373369](https://github.com/Azure/sonic-buildimage/pull/4529#discussion_r420373369) [](ancestors = 420373369,419745538)

Reply:
Let me investigate further.

Reply:
I tried the patch mentioned in https://sourceforge.net/p/net-snmp/bugs/2792/ and that didnt work.

```
/bin/bash ../libtool  --mode=link gcc -DNETSNMP_ENABLE_IPV6 -fno-strict-aliasing -g -O2 -fdebug-prefix-map=/sonic/src/snmpd/net-snmp-5.7.3+dfsg=. -specs=/usr/share/dpkg/no-pie-compile.specs -fstack-protector-strong -Wformat -Werror=format-security -DNETSNMP_USE_INLINE -Ulinux -Dlinux=linux  -D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fwrapv -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64  -I/usr/lib/x86_64-linux-gnu/perl/5.28/CORE   -o snmptrapd snmptrapd.lo  libnetsnmptrapd.la -specs=/usr/share/dpkg/no-pie-link.specs -Wl,-z,relro -Wl,-z,now  ../agent/libnetsnmpmibs.la ../agent/libnetsnmpagent.la ../snmplib/libnetsnmp.la  -L/usr/lib/x86_64-linux-gnu/ -lmariadb  -lnl-3 -lm  -Wl,-E -lwrap
libtool: link: gcc -DNETSNMP_ENABLE_IPV6 -fno-strict-aliasing -g -O2 -fdebug-prefix-map=/sonic/src/snmpd/net-snmp-5.7.3+dfsg=. -specs=/usr/share/dpkg/no-pie-compile.specs -fstack-protector-strong -Wformat -Werror=format-security -DNETSNMP_USE_INLINE -Ulinux -Dlinux=linux -D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fwrapv -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/lib/x86_64-linux-gnu/perl/5.28/CORE -o .libs/snmptrapd .libs/snmptrapd.o -specs=/usr/share/dpkg/no-pie-link.specs -Wl,-z -Wl,relro -Wl,-z -Wl,now -Wl,-E  ./.libs/libnetsnmptrapd.so ../agent/.libs/libnetsnmpmibs.so ../agent/.libs/libnetsnmpagent.so ../snmplib/.libs/libnetsnmp.so -L/usr/lib/x86_64-linux-gnu/ -lmariadb -lnl-3 -lm -lwrap 
/usr/bin/ld: ./.libs/libnetsnmptrapd.so: undefined reference to `load_defaults'
/usr/bin/ld: ./.libs/libnetsnmptrapd.so: undefined reference to `my_init'
collect2: error: ld returned 1 exit status
make[3]: *** [Makefile:297: snmptrapd] Error 1
make[3]: Leaving directory '/sonic/src/snmpd/net-snmp-5.7.3+dfsg/apps'
make[2]: *** [Makefile:657: subdirs] Error 1
make[2]: Leaving directory '/sonic/src/snmpd/net-snmp-5.7.3+dfsg'
dh_auto_build: make -j1 returned exit code 2
make[1]: *** [debian/rules:65: override_dh_auto_build] Error 2
make[1]: Leaving directory '/sonic/src/snmpd/net-snmp-5.7.3+dfsg'
make: *** [debian/rules:41: binary] Error 2
jjoseph@887ae551e879:/sonic/src/snmpd/net-snmp-5.7.3+dfsg$ dpkg -l | grep maria
ii  libmariadb-dev                         1:10.3.22-0+deb10u1             amd64        MariaDB database development files
ii  libmariadb-dev-compat:amd64            1:10.3.22-0+deb10u1             amd64        MariaDB Connector/C, compatibility symlinks
ii  libmariadb3:amd64                      1:10.3.22-0+deb10u1             amd64        MariaDB database client library
iU  libmariadbclient18:amd64               10.1.44-0+deb9u1                amd64        MariaDB database client library
ii  mariadb-common                         1:10.3.22-0+deb10u1             all          MariaDB common metapackage
jjoseph@887ae551e879:/sonic/src/snmpd/net-snmp-5.7.3+dfsg$
```
Even though we are explicitly linking against libmariadb, we still see the failures. The packages are available and installed too.

The patch @https://sourceforge.net/p/net-snmp/bugs/2782/#cbed is just wrapping the code within a #if block that is never satisfied.

Do you want me to change the code similar to the above patch?